// apropos/lib/adt/maybe.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Just = Just;
exports.of = of;
exports.Nothing = Nothing;
exports.empty = empty;
exports.validate = validate;
exports.fromNullable = fromNullable;

var _base = require('../base');

var _base2 = _interopRequireDefault(_base);

require('../index.h');

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class MaybeBase extends _base2.default {}

class MaybeJust extends MaybeBase {
  constructor(value) {
    super();
    this.value = value;
    /*:: return this */
  }

  ap(maybe) {
    return maybe.chain(fn => this.map(fn));
  }

  map(fn) {
    return new MaybeJust(fn(this.value));
  }

  chain(fn) {
    return fn(this.value);
  }

  tap(fn) {
    fn(this.value);
    return this;
  }

  fold(l, r) {
    return r(this.value);
  }

  orElse() {
    return this.value;
  }

  chainCond(cond, pass) {
    if (cond(this.value)) return new MaybeJust(pass(this.value));

    return new MaybeNothing();
  }

  logic(transformer) {
    var { cond, pass } = transformer;
    return this.chainCond(cond, pass);
  }

  pred(check) {
    if (check(this.value)) return this;

    return new MaybeNothing();
  }

  match(transformer) {
    var { Just } = transformer;
    return Just(this.value);
  }

  promise() {
    return Promise.resolve(this.value);
  }

  equals(value) {
    if (value instanceof MaybeJust) {
      return this.value === value.value;
    }
    return this.value === value;
  }

  alt() {
    return (/*:: addS( */this
    ); /*:: ) */
  }
  both(maybe) {
    if (maybe instanceof MaybeJust) {
      return new MaybeJust([this.value, maybe.value]);
    }

    return new MaybeNothing();
  }
  and(maybe) {
    return maybe;
  }

  isJust() {
    return true;
  }
  isNothing() {
    return false;
  }

  toString() {
    return 'Just( ' + (0, _util.failSafeStringify)(this.value) + ' )';
  }
}

class MaybeNothing extends MaybeBase {
  constructor() {
    super();
    /*:: return this */
  }

  ap() {
    return (/*:: changeT( */this
    ); /*:: ) */
  }

  map() {
    return (/*:: changeT( */this
    ); /*:: ) */
  }

  chain() {
    return (/*:: changeT( */this
    ); /*:: ) */
  }

  tap() {
    return this;
  }

  fold(l /*:: , r: (x: T) => O */) {
    return l();
  }

  orElse(x) {
    return x;
  }

  chainCond() {
    return new MaybeNothing();
  }

  logic(transformer) {
    var { cond, pass } = transformer;
    return this.chainCond(cond, pass);
  }

  pred() {
    return this;
  }

  match(transformer) {
    var { Nothing } = transformer;
    return Nothing();
  }

  promise() {
    return Promise.reject();
  }

  equals(value) {
    return value instanceof MaybeNothing;
  }

  alt(maybe) {
    return (/*:: addS( */maybe
    ); /*:: ) */
  }
  both() {
    return new MaybeNothing();
  }
  and() {
    return new MaybeNothing();
  }

  isJust() {
    return false;
  }
  isNothing() {
    return true;
  }

  toString() {
    return 'Nothing()';
  }
}

Object.defineProperty(MaybeJust.prototype, 'isomer', {
  value: 'just',
  enumerable: true,
  writable: false
});

Object.defineProperty(MaybeNothing.prototype, 'isomer', {
  value: 'nothing',
  enumerable: true,
  writable: false
});

(0, _util.addInterop)(MaybeJust, 'Maybe');
(0, _util.addInterop)(MaybeNothing, 'Maybe');

function Just(value) {
  return new MaybeJust(value);
}

function of(value) {
  return new MaybeJust(value);
}

function Nothing() {
  return new MaybeNothing();
}

function empty() {
  return new MaybeNothing();
}

function validate(validators) {
  return x => validators.reduce((acc, fn) => acc.pred(fn), new MaybeJust(x));
}

function fromNullable(value) {
  if (value == null) return new MaybeNothing();
  return new MaybeJust(value);
}
