// apropos/lib/adt/tuple.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.of = of;
exports.fst = fst;
exports.snd = snd;

var _base = require('../base');

var _base2 = _interopRequireDefault(_base);

require('../index.h');

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Tupleʹ extends _base2.default {
  constructor(a, b) {
    super();
    this /*::; noop */[0] = a;
    this /*::; noop */[1] = b;
    /*:: return this */
  }
  fst() {
    return this /*::, noop */[0];
  }
  snd() {
    return this /*::, noop */[1];
  }
  bimap(f, g) {
    return new Tupleʹ(f(this.fst()), g(this.snd()));
  }
  map(f) {
    return new Tupleʹ(this.fst(), f(this.snd()));
  }
  curry(f) {
    return f(this);
  }
  uncurry(f) {
    return f(this.fst(), this.snd());
  }
  extend(f) {
    return new Tupleʹ(this.fst(), f(this));
  }
  extract() {
    return this.snd();
  }
  foldl(f, z) {
    return f(this.snd(), z);
  }
  foldr(f, z) {
    return f(z, this.snd());
  }
  foldMap(f) {
    return f(this.snd());
  }
  equals(tuple) {
    if (tuple instanceof Tupleʹ) {
      return this.fst() === tuple.fst() && this.snd() === tuple.snd();
    }
    return false;
    // this.eqFirst(tuple)
    //   && this.eqSecond(tuple)
  }
  // eqFirst<Aʹ, -Bʹ>(tuple: Tuple<Aʹ, Bʹ>): boolean {
  //   return eq(this.fst(), tuple.fst())
  // }
  // eqSecond<-Aʹ, Bʹ>(tuple: Tuple<Aʹ, Bʹ>): boolean {
  //   return eq(this.snd(), tuple.snd())
  // }
  toValue() {
    return [this.fst(), this.snd()];
  }
  toJSON() {
    return this.toValue();
  }
  toString() {
    return `Tuple( ${(0, _util.failSafeStringify)(this.fst())}, ${(0, _util.failSafeStringify)(this.snd())} )`;
  }
  // static of<Aʹ, Bʹ>(a: Aʹ, b: Bʹ): Tuple<Aʹ, Bʹ> {
  //   return new Tupleʹ(a, b)
  // }
}

/* eslint-disable no-whitespace-before-property */

(0, _util.addInterop)(Tupleʹ, 'Tuple');

/*::
const noop = {}
*/

var additionalProps = {
  length: {
    value: 2
  }
};

Object.defineProperties(Tupleʹ.prototype, additionalProps);

function of(a, b) {
  return new Tupleʹ(a, b);
}

function fst(tuple) {
  return tuple.fst();
}

function snd(tuple) {
  return tuple.snd();
}
