// apropos/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tuple = exports.Identity = exports.Maybe = exports.Either = exports.makeError = exports.fold = exports.ofL = exports.of = exports.Left = exports.Right = exports.is = undefined;

var _either = require('./adt/either');

Object.defineProperty(exports, 'is', {
  enumerable: true,
  get: function () {
    return _either.is;
  }
});
Object.defineProperty(exports, 'Right', {
  enumerable: true,
  get: function () {
    return _either.Right;
  }
});
Object.defineProperty(exports, 'Left', {
  enumerable: true,
  get: function () {
    return _either.Left;
  }
});
Object.defineProperty(exports, 'of', {
  enumerable: true,
  get: function () {
    return _either.of;
  }
});
Object.defineProperty(exports, 'ofL', {
  enumerable: true,
  get: function () {
    return _either.ofL;
  }
});

var _pointFree = require('./point-free');

Object.defineProperty(exports, 'fold', {
  enumerable: true,
  get: function () {
    return _pointFree.fold;
  }
});

var _error = require('./error');

Object.defineProperty(exports, 'makeError', {
  enumerable: true,
  get: function () {
    return _error.makeError;
  }
});

require('./index.h');

require('./error');

var _Either = _interopRequireWildcard(_either);

var _maybe = require('./adt/maybe');

var _Maybe = _interopRequireWildcard(_maybe);

var _identity = require('./adt/identity');

var _Identity = _interopRequireWildcard(_identity);

var _tuple = require('./adt/tuple');

var _Tuple = _interopRequireWildcard(_tuple);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

exports.default = _either.of;
exports.Either = _Either;
exports.Maybe = _Maybe;
exports.Identity = _Identity;
exports.Tuple = _Tuple;
